cmd/endpoint-ellipses.go

```diff
@@ -0,0 +1,251 @@
+/*
+ * Minio Cloud Storage, (C) 2018 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/minio/minio-go/pkg/set"
+	"github.com/minio/minio/pkg/ellipses"
+)
+
+// This file implements and supports ellipses pattern for
+// `minio server` command line arguments.
+
+// Maximum number of unique args supported on the command line.
+const (
+	serverCommandLineArgsMax = 32
+)
+
+// Endpoint set represents parsed ellipses values, also provides
+// methods to get the sets of endpoints.
+type endpointSet struct {
+	argPatterns []ellipses.ArgPattern
+	endpoints   []string   // Endpoints saved from previous GetEndpoints().
+	setIndexes  [][]uint64 // All the sets.
+}
+
+// Supported set sizes this is used to find the optimal
+// single set size.
+var setSizes = []uint64{4, 6, 8, 10, 12, 14, 16}
+
+// getDivisibleSize - returns a greatest common divisor of
+// all the ellipses sizes.
+func getDivisibleSize(totalSizes []uint64) (result uint64) {
+	gcd := func(x, y uint64) uint64 {
+		for y != 0 {
+			x, y = y, x%y
+		}
+		return x
+	}
+	if len(totalSizes) > 1 {
+		result = gcd(totalSizes[0], totalSizes[1])
+		totalSizes = totalSizes[2:]
+		for i := 0; i < len(totalSizes); i++ {
+			result = gcd(result, totalSizes[i])
+		}
+		return result
```

Comment:
More efficient and clean:

```go
result = totalSizes[0]
for i := 1; i < len(totalSizes); i++ {
    result = gcd(result, totalSizes[i])
}
return result
```

with the added advantage that it works for `len(totalSizes) = 1` too.